aws/src/main/java/org/apache/iceberg/aws/RESTSigV4Signer.java

```diff
@@ -0,0 +1,112 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.aws;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.hc.core5.http.EntityDetails;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.HttpRequest;
+import org.apache.hc.core5.http.HttpRequestInterceptor;
+import org.apache.hc.core5.http.io.entity.StringEntity;
+import org.apache.hc.core5.http.protocol.HttpContext;
+import org.apache.iceberg.exceptions.RESTException;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.signer.Aws4Signer;
+import software.amazon.awssdk.auth.signer.internal.SignerConstant;
+import software.amazon.awssdk.auth.signer.params.Aws4SignerParams;
+import software.amazon.awssdk.auth.signer.params.SignerChecksumParams;
+import software.amazon.awssdk.core.checksums.Algorithm;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpMethod;
+import software.amazon.awssdk.regions.Region;
+
+public class RESTSigV4Signer implements HttpRequestInterceptor {
+  private static final String EMPTY_BODY_SHA256 =
+      "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
+
+  private final Aws4Signer signer = Aws4Signer.create();
+  private AwsCredentialsProvider credentialsProvider;
+
+  private String signingName;
+  private Region signingRegion;
+
+  public void initialize(Map<String, String> properties) {
+    AwsProperties awsProperties = new AwsProperties(properties);
+
+    this.signingRegion = awsProperties.restSigv4SigningRegion();
+    this.signingName = awsProperties.restSigv4SigningName();
+    this.credentialsProvider = awsProperties.restSigv4CredentialsProvider();
+  }
+
+  @Override
+  public void process(HttpRequest request, EntityDetails entity, HttpContext context) {
+    try {
+      Aws4SignerParams params =
+          Aws4SignerParams.builder()
+              .signingName(signingName)
+              .signingRegion(signingRegion)
+              .awsCredentials(credentialsProvider.resolveCredentials())
+              .checksumParams(
+                  SignerChecksumParams.builder()
+                      .algorithm(Algorithm.SHA256)
+                      .isStreamingRequest(false)
```

Comment:
There is a `request.getEntity().isStreaming()` method from Apache. Should we add a precondition that verifies this is true to be safe?

Reply:
The only entity we support is `StringEntity` which is not streaming by definition.  We now throw if it's a different entity type because we don't use or support anything else at this point.  The new case makes this check somewhat redundant.